nengo/spa/vocab.py

```diff
@@ -21,149 +26,109 @@ class Vocabulary(object):
         is not generated after the specified number of attempts, the
         candidate pointer with lowest maximum cosine with all existing
         pointers is returned.
         """
-        if self.randomize:
-            if self.vectors.shape[0] == 0:
+        if len(self) == 0:
+            best_p = pointer.SemanticPointer(self.dimensions, rng=self.rng)
+        else:
+            best_p = None
+            best_sim = np.inf
+            for _ in range(attempts):
                 p = pointer.SemanticPointer(self.dimensions, rng=self.rng)
+                if transform is not None:
+                    p = eval('p.' + transform, {'p': p}, self)
+                p_sim = np.max(np.dot(self._vectors, p.v))
+                if p_sim < best_sim:
+                    best_p = p
+                    best_sim = p_sim
+                    if p_sim < self.max_similarity:
+                        break
             else:
-                p_sim = np.inf
-                for _ in range(attempts):
-                    pp = pointer.SemanticPointer(self.dimensions, rng=self.rng)
-                    pp_sim = max(np.dot(self.vectors, pp.v))
-                    if pp_sim < p_sim:
-                        p = pp
-                        p_sim = pp_sim
-                        if p_sim < self.max_similarity:
-                            break
-                else:
-                    warnings.warn(
-                        'Could not create a semantic pointer with '
-                        'max_similarity=%1.2f (D=%d, M=%d)'
-                        % (self.max_similarity,
-                           self.dimensions,
-                           len(self.pointers)))
-
-            # Check and make vector unitary if needed
-            if unitary:
-                p.make_unitary()
-        else:
-            index = len(self.pointers)
-            if index >= self.dimensions:
-                raise ValidationError(
-                    "Tried to make more semantic pointers than "
-                    "dimensions with non-randomized Vocabulary",
-                    attr='dimensions', obj=self)
-            p = pointer.SemanticPointer(np.eye(self.dimensions)[index])
-        return p
+                warnings.warn(
+                    'Could not create a semantic pointer with '
+                    'max_similarity=%1.2f (D=%d, M=%d)'
+                    % (self.max_similarity, self.dimensions,
+                       len(self.pointers)))
+        return best_p
+
+    def __contains__(self, key):
+        return key in self.pointers
+
+    def __len__(self):
+        return len(self._vectors)
+
+    def __iter__(self):
+        return iter(self._keys)
 
     def __getitem__(self, key):
-        """Return the semantic pointer with the requested name.
+        """Return the semantic pointer with the requested name."""
+        if not self.strict and key not in self:
+            assert '.' not in key
+            assert ';' not in key
```

Comment:
`vocab['A+B']` will actually work. It might not be super useful because the only way to access the SP `'A+B'` is via the `__getitem__` method; it cannot be used in parsed expressions, but it doesn't break anything. Names containing `'.'` or `';'` (and looking at the code also `'='` which should probably be added) will lead to a `KeyError` because `populate` will split the name at those symbols, but `__getitem__` looks for the full name. I think I inserted the asserts to fail early before `populate` already modified the vocab in an unintended way.